lib/componentmanager.js

```diff
@@ -429,28 +473,56 @@
      * 
      * @see ComponentManager.prototype.addComponent
      * @param {Layer} layer The layer from which to extract components
-     * @return {Array.<{errors: Array.<string>} | {component: Component}>}
+     * @return {Promise} resolves to Array.<{errors: Array.<string>} | {component: Component}>}
      */
     ComponentManager.prototype.findAllComponents = function (layer) {
         
-        var results,
-            components = [];
-
-        if (layer.name) {
-            results = this._parserManager.analyzeLayerName(layer.name);
-            results.forEach(function (result) {
-                var errors = result.errors;
-                if (errors.length > 0) {
-                    components.push({errors: errors});
-                } else {
-                    var component = result.component;
-                    if (component.file || component.default) {
-                        components.push({component: component});
-                    }
+        var findAllDeferred = Q.defer(),
+            _G = this._generator,
+            _PM = this._parserManager,
+            doc = this._document,
+            fnFromLayerName = function (layerName) {
+                var aRet = [],
+                    results;
+                if (layerName) {
+                    results = _PM.analyzeLayerName(layerName);
+                    results.forEach(function (result) {
+                        var errors = result.errors;
+                        if (errors.length > 0) {
+                            aRet.push({errors: errors});
+                        } else {
+                            var component = result.component;
+                            if (component.file || component.default) {
+                                aRet.push({component: component});
+                            }
+                        }
+                    });
                 }
-            }, this);
+                return aRet;
+            };
+        
+        if (!this._config["meta-data-driven"]) {
+            findAllDeferred.resolve(fnFromLayerName(layer.name));
+        } else {
+            _G.getDocumentSettingsForPlugin(doc.id, META_PLUGIN_ID).then(function (docMeta) {
+                if (docMeta && docMeta.metaEnabled) {
+                    _G.getLayerSettingsForPlugin(doc.id, layer.id, META_PLUGIN_ID).then(function (layerMeta) {
```

Comment:
Can we get these settings with getDocumentInfo or the change notifications? If they were already on the layer and up to date this would be faster and not require the promises